{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"batch-correction-download example.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyO+to1DPMQ6wcWw/mpEmjQQ"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"zqgQNveeYcsX","executionInfo":{"status":"ok","timestamp":1644391715550,"user_tz":-480,"elapsed":1400,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"16f7910e-b6f6-485f-f74b-e424e1470033"},"outputs":[{"output_type":"stream","name":"stdout","text":["--2022-02-09 07:28:33--  http://genomedata.org/rnaseq-tutorial/batch_correction/GSE48035_ILMN.counts.txt.gz\n","Resolving genomedata.org (genomedata.org)... 54.71.55.4\n","Connecting to genomedata.org (genomedata.org)|54.71.55.4|:80... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 3260867 (3.1M) [application/x-gzip]\n","Saving to: ‘GSE48035_ILMN.counts.txt.gz’\n","\n","GSE48035_ILMN.count 100%[===================>]   3.11M  4.28MB/s    in 0.7s    \n","\n","2022-02-09 07:28:34 (4.28 MB/s) - ‘GSE48035_ILMN.counts.txt.gz’ saved [3260867/3260867]\n","\n"]}],"source":["!wget http://genomedata.org/rnaseq-tutorial/batch_correction/GSE48035_ILMN.counts.txt.gz\n"]},{"cell_type":"code","source":["# remove all quotes from file\n","!zcat GSE48035_ILMN.counts.txt.gz | tr -d '\"' > GSE48035_ILMN.counts.tmp.txt\n","!ls"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ax7EIApVYg0n","executionInfo":{"status":"ok","timestamp":1644391757749,"user_tz":-480,"elapsed":530,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"34e6a3ff-7ba3-4e04-ae59-093602f21273"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["GSE48035_ILMN.counts.tmp.txt  GSE48035_ILMN.counts.txt.gz  sample_data\n"]}]},{"cell_type":"code","source":["%%bash\n","#create a fixed version of the header and store for later\n","head -n 1 GSE48035_ILMN.counts.tmp.txt | perl -ne 'print \"Gene\\tChr\\t$_\"' > header.txt\n","\n","#split the chromosome and gene names on each line, sort the file by gene name\n","perl -ne 'chomp; if ($_ =~ /^(chr\\w+)\\!(\\S+)(.*)/){print \"$2\\t$1$3\\n\"}else{print \"$_\\n\"}' GSE48035_ILMN.counts.tmp.txt | sort > GSE48035_ILMN.counts.tmp2.txt\n","\n","#remove the old header line\n","grep -v --color=never ABRF GSE48035_ILMN.counts.tmp2.txt > GSE48035_ILMN.counts.clean.txt\n","\n","#cut out columns for the UHR (A) and HBR (B) samples, replicates 1-4, and PolyA vs Enrichment \n","cut -f 1-2,3-6,7-10,19-22,23-26 GSE48035_ILMN.counts.clean.txt > GSE48035_ILMN.Counts.SampleSubset.txt\n","cut -f 1-2,3-6,7-10,19-22,23-26 header.txt > header.SampleSubset.txt\n","\n","#how many gene lines are we starting with?\n","wc -l GSE48035_ILMN.Counts.SampleSubset.txt\n","\n","#cleanup \n","rm -f GSE48035_ILMN.counts.txt.gz GSE48035_ILMN.counts.tmp.txt GSE48035_ILMN.counts.tmp2.txt GSE48035_ILMN.counts.clean.txt header.txt"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IQ6VcxbTYpFZ","executionInfo":{"status":"ok","timestamp":1644391762598,"user_tz":-480,"elapsed":533,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"f4c56cad-53fa-4d92-8717-acdc3c44f177"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["35779 GSE48035_ILMN.Counts.SampleSubset.txt\n"]}]},{"cell_type":"markdown","source":["Limit these counts to those that correspond to known protein coding genes"],"metadata":{"id":"2ILqYkTFYwRP"}},{"cell_type":"code","source":["%%bash\n","#download complete Ensembl GTF file\n","wget ftp://ftp.ensembl.org/pub/release-101/gtf/homo_sapiens/Homo_sapiens.GRCh38.101.gtf.gz\n","\n","#grab all the gene records, limit to gene with \"protein_coding\" biotype, create unique gene name list\n","zcat Homo_sapiens.GRCh38.101.gtf.gz | grep -w gene | grep \"gene_biotype \\\"protein_coding\\\"\" | cut -f 9 | cut -d \";\" -f 3 | tr -d \" gene_name \" | tr -d '\"' | sort | uniq > Ensembl101_ProteinCodingGeneNames.txt\n","\n","#how many unique protein coding genes names does Ensembl have?\n","wc -l Ensembl101_ProteinCodingGeneNames.txt\n","\n","#filter our gene count matrix down to only the protein coding genes\n","join -j 1 -t $'\\t' Ensembl101_ProteinCodingGeneNames.txt GSE48035_ILMN.Counts.SampleSubset.txt | cat header.SampleSubset.txt - > GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv\n","\n","#how many lines of RNA-seq count do we still have?\n","wc -l GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv\n","\n","#clean up \n","rm -f header.SampleSubset.txt GSE48035_ILMN.Counts.SampleSubset.txt\n","\n","#take a look at the final filtered read count matrix to be used for the following analysis\n","column -t GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv | less -S"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"6QnOBWFeY3tT","executionInfo":{"status":"ok","timestamp":1644391836359,"user_tz":-480,"elapsed":14651,"user":{"displayName":"Indrik Wijaya","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"00208067985282844199"}},"outputId":"6aa41846-b499-4917-933f-e450298ffb7d"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stderr","text":["IOPub data rate exceeded.\n","The notebook server will temporarily stop sending output\n","to the client in order to avoid crashing it.\n","To change this limit, set the config variable\n","`--NotebookApp.iopub_data_rate_limit`.\n","\n","Current values:\n","NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n","NotebookApp.rate_limit_window=3.0 (secs)\n","\n"]}]},{"cell_type":"code","source":[""],"metadata":{"id":"GxOd3TjJY-1H"},"execution_count":null,"outputs":[]}]}