{"metadata":{"kernelspec":{"name":"ir","display_name":"R","language":"R"},"language_info":{"name":"R","codemirror_mode":"r","pygments_lexer":"r","mimetype":"text/x-r-source","file_extension":".r","version":"4.0.5"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This R environment comes with many helpful analytics packages installed\n# It is defined by the kaggle/rstats Docker image: https://github.com/kaggle/docker-rstats\n# For example, here's a helpful package to load\n\nlibrary(tidyverse) # metapackage of all tidyverse packages\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nlist.files(path = \"../input\")\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"051d70d956493feee0c6d64651c6a088724dca2a","_execution_state":"idle","execution":{"iopub.status.busy":"2022-02-21T07:17:24.378991Z","iopub.execute_input":"2022-02-21T07:17:24.381275Z","iopub.status.idle":"2022-02-21T07:17:25.573294Z"},"trusted":true},"execution_count":1,"outputs":[]},{"cell_type":"markdown","source":"Source:\nhttps://rnabio.org/module-03-expression/0003/05/01/Batch-Correction/\n\nThis publication used the same UHR (cancer cell lines) and HBR (brain tissue) samples we have been using throughout this course. To examine a strong batch effect, we will consider a DE analysis of UHR vs HBR where we compare Ribo-depleted (“Ribo”) and polyA-enriched (“Poly”) samples.\n\nIf we do DE analysis of UHR vs. HBR for replicates that are consistent with respect to Ribo-depletion or PolyA-enrichment, how does the result compare to when we use Ribo-depleted data vs PolyA-enriched data? If you do batch correction and redo these comparisons does it make the results more comparable? i.e. can we correct for the technical differences introduced by the library construction approach and see the same biological differences?\n\nThis is a bit contrived because there really are true biological differences expected for polyA vs. Ribo-depleted data. To counter this, we will limit the analysis to only know protein coding genes. For those genes we expect essentially the same biological answer when comparing UHR vs HBR expression patterns.\n\nThis exercise is also a bit simplistic in the sense that we have perfectly balanced conditions and batches. Our conditions of interest are: HBR (brain) vs. UHR (cancer cell line) expression patterns. Our batches are the two methods of processing: Riboreduction and PolyA enrichment. And we have 4 replicates of both conditions in both batches. To perform this kind of batch correction you need at least some representation of each of your conditions of interest in each batch. So, for example, if we processed all the HBR samples with Riboreduction and all the UHR samples with PolyA enrichment, we would be unable to model the batch effect vs the condition effect.","metadata":{}},{"cell_type":"code","source":"BiocManager::install(c('sva', 'edgeR'))\nlibrary(sva)\nlibrary(ggplot2)\nlibrary(gridExtra)\nlibrary(edgeR)\nlibrary(UpSetR)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:17:25.576502Z","iopub.execute_input":"2022-02-21T07:17:25.604368Z","iopub.status.idle":"2022-02-21T07:24:00.920934Z"},"trusted":true},"execution_count":2,"outputs":[]},{"cell_type":"code","source":"#load in the uncorrected data as raw counts\nuncorrected_data = read.table(\"/kaggle/input/batcheffectclean/GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv\", \n                              header=TRUE, sep=\"\\t\", as.is=c(1,2))\n\n#simplify the names of the data columns\n# (A = Universal Human Reference RNA and B = Human Brain Reference RNA)\n# RNA = polyA enrichment and RIBO = ribosomal RNA depletion\n# 1, 2, 3, 4 are replicates\nnames(uncorrected_data) = c(\"Gene\", \"Chr\", \"UHR_Ribo_1\", \"UHR_Ribo_2\", \"UHR_Ribo_3\", \"UHR_Ribo_4\", \"HBR_Ribo_1\", \"HBR_Ribo_2\", \"HBR_Ribo_3\", \"HBR_Ribo_4\", \n                            \"UHR_Poly_1\", \"UHR_Poly_2\", \"UHR_Poly_3\", \"UHR_Poly_4\", \"HBR_Poly_1\", \"HBR_Poly_2\", \"HBR_Poly_3\", \"HBR_Poly_4\")\nsample_names = names(uncorrected_data)[3:length(names(uncorrected_data))]\n\n#review data structure\nhead(uncorrected_data)\ndim(uncorrected_data)\n\n#define conditions, library methods, and replicates\nconditions = c(\"UHR\", \"UHR\", \"UHR\", \"UHR\", \"HBR\", \"HBR\", \"HBR\", \"HBR\", \"UHR\", \"UHR\", \"UHR\", \"UHR\", \"HBR\", \"HBR\", \"HBR\", \"HBR\")\nlibrary_methods = c(\"Ribo\", \"Ribo\", \"Ribo\", \"Ribo\", \"Ribo\", \"Ribo\", \"Ribo\", \"Ribo\", \"Poly\", \"Poly\", \"Poly\", \"Poly\", \"Poly\", \"Poly\", \"Poly\", \"Poly\")\nreplicates = c(1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4)\n\n#calculate principal components for the uncorrected data\npca_uncorrected_obj = prcomp(uncorrected_data[,sample_names])\n\n#pull PCA values out of the PCA object\npca_uncorrected = as.data.frame(pca_uncorrected_obj[2]$rotation)\n\n#assign labels to the data frame\npca_uncorrected[,\"condition\"] = conditions\npca_uncorrected[,\"library_method\"] = library_methods\npca_uncorrected[,\"replicate\"] = replicates\n\n#plot the PCA\n#create a classic 2-dimension PCA plot (first two principal components) with conditions and library methods indicated\ncols <- c(\"UHR\" = \"#481567FF\", \"HBR\" = \"#1F968BFF\")\np1 = ggplot(data=pca_uncorrected, aes(x=PC1, y=PC2, color=condition, shape=library_method))\np1 = p1 + geom_point(size=3)\np1 = p1 + stat_ellipse(type=\"norm\", linetype=2)\np1 = p1 + labs(title=\"PCA, RNA-seq counts for 16 HBR/UHR and Ribo/PolyA samples (uncorrected data)\", color=\"Condition\", shape=\"Library Method\")\np1 = p1 + scale_colour_manual(values = cols)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:00.924109Z","iopub.execute_input":"2022-02-21T07:24:00.926504Z","iopub.status.idle":"2022-02-21T07:24:01.230962Z"},"trusted":true},"execution_count":3,"outputs":[]},{"cell_type":"code","source":"p1","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:01.245037Z","iopub.execute_input":"2022-02-21T07:24:01.247555Z","iopub.status.idle":"2022-02-21T07:24:01.967526Z"},"trusted":true},"execution_count":4,"outputs":[]},{"cell_type":"code","source":"# Batch correction\n#first we need to transform the format of our groups and batches from names (e.g. \"UHR\", \"HBR\", etc.) to numbers (e.g. 1, 2, etc.)\n#in the command below \"sapply\" is used to apply the \"switch\" command to each element and convert names to numbers as we define\ngroups = sapply(as.character(conditions), switch, \"UHR\" = 1, \"HBR\" = 2, USE.NAMES = F)\nbatches = sapply(as.character(library_methods), switch, \"Ribo\" = 1, \"Poly\" = 2, USE.NAMES = F)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:01.969744Z","iopub.execute_input":"2022-02-21T07:24:01.971102Z","iopub.status.idle":"2022-02-21T07:24:01.985246Z"},"trusted":true},"execution_count":5,"outputs":[]},{"cell_type":"code","source":"groups","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:01.987448Z","iopub.execute_input":"2022-02-21T07:24:01.988789Z","iopub.status.idle":"2022-02-21T07:24:02.004056Z"},"trusted":true},"execution_count":6,"outputs":[]},{"cell_type":"code","source":"batches","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:02.006284Z","iopub.execute_input":"2022-02-21T07:24:02.007632Z","iopub.status.idle":"2022-02-21T07:24:02.022486Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"code","source":"corrected_data = ComBat_seq(counts = as.matrix(uncorrected_data[, sample_names]),\n                           batch = batches, group = groups)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:02.024775Z","iopub.execute_input":"2022-02-21T07:24:02.026113Z","iopub.status.idle":"2022-02-21T07:24:12.673434Z"},"trusted":true},"execution_count":8,"outputs":[]},{"cell_type":"code","source":"#join the gene and chromosome names onto the now corrected counts from ComBat_seq\ncorrected_data = cbind(uncorrected_data[,c(\"Gene\",\"Chr\")], corrected_data)\n\n#compare dimensions of corrected and uncorrected data sets\ndim(uncorrected_data)\ndim(corrected_data)\n\n#visually compare values of corrected and uncorrected data sets\nhead(uncorrected_data)\nhead(corrected_data)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:12.676793Z","iopub.execute_input":"2022-02-21T07:24:12.678267Z","iopub.status.idle":"2022-02-21T07:24:12.749521Z"},"trusted":true},"execution_count":9,"outputs":[]},{"cell_type":"markdown","source":"## Perform PCA analysis on the batch corrected data & contrast with the uncorrected data","metadata":{}},{"cell_type":"code","source":"#calculate principal components for the uncorrected data\npca_corrected_obj = prcomp(corrected_data[,sample_names])\n\n#pull PCA values out of the PCA object\npca_corrected = as.data.frame(pca_corrected_obj[2]$rotation)\n\n#assign labels to the data frame\npca_corrected[,\"condition\"] = conditions\npca_corrected[,\"library_method\"] = library_methods\npca_corrected[,\"replicate\"] = replicates\n\n#as above, create a PCA plot for comparison to the uncorrected data\ncols <- c(\"UHR\" = \"#481567FF\", \"HBR\" = \"#1F968BFF\")\np2 = ggplot(data=pca_corrected, aes(x=PC1, y=PC2, color=condition, shape=library_method))\np2 = p2 + geom_point(size=3)\np2 = p2 + stat_ellipse(type=\"norm\", linetype=2)\np2 = p2 + labs(title=\"PCA, RNA-seq counts for 16 HBR/UHR and Ribo/PolyA samples (batch corrected data)\", color=\"Condition\", shape=\"Library Method\")\np2 = p2 + scale_colour_manual(values = cols)\n\npdf(file=\"Uncorrected-vs-BatchCorrected-PCA.pdf\")\ngrid.arrange(p1, p2, nrow = 2)\ndev.off()","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:12.752954Z","iopub.execute_input":"2022-02-21T07:24:12.754381Z","iopub.status.idle":"2022-02-21T07:24:13.314550Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"code","source":"grid.arrange(p1, p2, nrow = 2)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:13.316817Z","iopub.execute_input":"2022-02-21T07:24:13.318178Z","iopub.status.idle":"2022-02-21T07:24:13.989310Z"},"trusted":true},"execution_count":11,"outputs":[]},{"cell_type":"markdown","source":"## Perform DE analysis of the corrected & uncorrected data\n\nHow does batch correction influence differential gene expression results? Use UpSet plots to examine the overlap of significant DE genes found for the following comparisons:\n\n* UHR-Ribo vs HBR-Ribo (same library type, 4 vs 4 replicates)\n* UHR-Poly vs HBR-Poly (same library type, 4 vs 4 replicates)\n* UHR-Ribo vs HBR-Poly (different library types, 4 vs 4 replicates)\n* UHR-Poly vs HBR-Ribo (different library types, 4 vs 4 replicates)\n* UHR-Comb vs HBR-Comb (combined library types, 8 vs 8 replicates)\n\nThese five differential expression analysis comparisons will be performed with both the uncorrected and corrected data.\n\n1. Does correction increase agreement between the five comparisons?\n2. Does it appear to increase statistical power when combining all 8 replicates of UHR and HBR?\n3. What do we expect to see for comparisons like UHR-Ribo vs HBR-Poly before and after batch correction? Do we expect correction to increase or decrease the number of significant results?","metadata":{}},{"cell_type":"code","source":"#perform differential expression analysis on the uncorrected data and batch corrected data sets\n\n#first define the sets of samples to be compared to each other\nuhr_ribo_samples = c(\"UHR_Ribo_1\", \"UHR_Ribo_2\", \"UHR_Ribo_3\", \"UHR_Ribo_4\")\nuhr_poly_samples = c(\"UHR_Poly_1\", \"UHR_Poly_2\", \"UHR_Poly_3\", \"UHR_Poly_4\")\nhbr_ribo_samples = c(\"HBR_Ribo_1\", \"HBR_Ribo_2\", \"HBR_Ribo_3\", \"HBR_Ribo_4\")\nhbr_poly_samples = c(\"HBR_Poly_1\", \"HBR_Poly_2\", \"HBR_Poly_3\", \"HBR_Poly_4\")\nuhr_samples = c(uhr_ribo_samples, uhr_poly_samples)\nhbr_samples = c(hbr_ribo_samples, hbr_poly_samples)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:13.991754Z","iopub.execute_input":"2022-02-21T07:24:13.993169Z","iopub.status.idle":"2022-02-21T07:24:14.026519Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"code","source":"#create a function that will run edgeR (DE analysis) for a particular pair of sample sets\nrun_edgeR = function(data, group_a_name, group_a_samples, group_b_samples, group_b_name){\n  #create a list of all samples for this current comparison\n  samples_for_comparison = c(group_a_samples, group_b_samples)\n  \n  #define the class factor for this pair of sample sets\n  class = factor(c(rep(group_a_name,length(group_a_samples)), rep(group_b_name,length(group_b_samples))))\n  \n  #create a simplified data matrix for only these samples\n  rawdata = data[,samples_for_comparison]\n  \n  #store gene names for later\n  genes = rownames(data)\n  gene_names = data[,\"Gene\"]\n  \n  #make DGElist object\n  y = DGEList(counts=rawdata, genes=genes, group=class)\n  \n  #perform TMM normalization\n  y <- calcNormFactors(y)\n  \n  #estimate dispersion\n  y <- estimateCommonDisp(y, verbose=TRUE)\n  y <- estimateTagwiseDisp(y)\n  \n  #differential expression test\n  et <- exactTest(y)\n  \n  #print number of up/down significant genes at FDR = 0.05 significance level and store the DE status in a new variable (de)\n  summary(de <- decideTestsDGE(et, p=.05))\n  print(summary(de <- decideTestsDGE(et, adjust.method=\"fdr\", p=.05)))\n  \n  #create a matrix of significant DE genes\n  mat <- cbind(\n    genes, gene_names,\n    sprintf('%0.3f', log10(et$table$PValue)),\n    sprintf('%0.3f', et$table$logFC)\n  )[as.logical(de),]\n  colnames(mat) <- c(\"Gene\", \"Gene_Name\", \"Log10_Pvalue\", \"Log_fold_change\")\n  \n  #order by log fold change\n  o <- order(et$table$logFC[as.logical(de)],decreasing=TRUE)\n  mat <- mat[o,]\n\n  #fix the issue where corrected p-values that are 0 become -Inf upon log10 conversion\n  x = as.numeric(mat[,\"Log10_Pvalue\"])\n  lowest_pvalue = min(x[which(!x == -Inf)])\n  i = which(mat[,\"Log10_Pvalue\"] == -Inf)\n  mat[i,\"Log10_Pvalue\"] = lowest_pvalue\n  \n  return(mat)\n}","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:14.028867Z","iopub.execute_input":"2022-02-21T07:24:14.030170Z","iopub.status.idle":"2022-02-21T07:24:14.044048Z"},"trusted":true},"execution_count":13,"outputs":[]},{"cell_type":"code","source":"#run the five comparisons through edgeR using the *uncorrected data*\nuhr_ribo_vs_hbr_ribo_uncorrected = run_edgeR(data=uncorrected_data, group_a_name=\"UHR\", group_a_samples=uhr_ribo_samples, group_b_name=\"HBR\", group_b_samples=hbr_ribo_samples)\nuhr_poly_vs_hbr_poly_uncorrected = run_edgeR(data=uncorrected_data, group_a_name=\"UHR\", group_a_samples=uhr_poly_samples, group_b_name=\"HBR\", group_b_samples=hbr_poly_samples)\nuhr_ribo_vs_hbr_poly_uncorrected = run_edgeR(data=uncorrected_data, group_a_name=\"UHR\", group_a_samples=uhr_ribo_samples, group_b_name=\"HBR\", group_b_samples=hbr_poly_samples)\nuhr_poly_vs_hbr_ribo_uncorrected = run_edgeR(data=uncorrected_data, group_a_name=\"UHR\", group_a_samples=uhr_poly_samples, group_b_name=\"HBR\", group_b_samples=hbr_ribo_samples)\nuhr_vs_hbr_uncorrected = run_edgeR(data=uncorrected_data, group_a_name=\"UHR\", group_a_samples=uhr_samples, group_b_name=\"HBR\", group_b_samples=hbr_samples)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:14.047360Z","iopub.execute_input":"2022-02-21T07:24:14.048812Z","iopub.status.idle":"2022-02-21T07:24:29.846797Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"code","source":"#run the same five comparisons through edgeR using the *batch corrected data*\nuhr_ribo_vs_hbr_ribo_corrected = run_edgeR(data=corrected_data, group_a_name=\"UHR\", group_a_samples=uhr_ribo_samples, group_b_name=\"HBR\", group_b_samples=hbr_ribo_samples)\nuhr_poly_vs_hbr_poly_corrected = run_edgeR(data=corrected_data, group_a_name=\"UHR\", group_a_samples=uhr_poly_samples, group_b_name=\"HBR\", group_b_samples=hbr_poly_samples)\nuhr_ribo_vs_hbr_poly_corrected = run_edgeR(data=corrected_data, group_a_name=\"UHR\", group_a_samples=uhr_ribo_samples, group_b_name=\"HBR\", group_b_samples=hbr_poly_samples)\nuhr_poly_vs_hbr_ribo_corrected = run_edgeR(data=corrected_data, group_a_name=\"UHR\", group_a_samples=uhr_poly_samples, group_b_name=\"HBR\", group_b_samples=hbr_ribo_samples)\nuhr_vs_hbr_corrected = run_edgeR(data=corrected_data, group_a_name=\"UHR\", group_a_samples=uhr_samples, group_b_name=\"HBR\", group_b_samples=hbr_samples)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:29.848853Z","iopub.execute_input":"2022-02-21T07:24:29.850108Z","iopub.status.idle":"2022-02-21T07:24:43.717408Z"},"trusted":true},"execution_count":15,"outputs":[]},{"cell_type":"code","source":"#how much of a difference does batch correction make when doing the comparison of all UHR vs all HBR samples?\ndim(uhr_vs_hbr_uncorrected)\ndim(uhr_vs_hbr_corrected)","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:43.719582Z","iopub.execute_input":"2022-02-21T07:24:43.720823Z","iopub.status.idle":"2022-02-21T07:24:43.740206Z"},"trusted":true},"execution_count":16,"outputs":[]},{"cell_type":"markdown","source":"What differs in each comparison is:\n\n* whether we are comparing replicates prepared with the same library construction approach (Ribo reduction or PolyA) or whether we are comparing data prepared with different approaches\n* whether we are comparing 4 UHR vs 4 HBR replicates according to these library construction approaches, or pooling all 8 UHR and all 8 HBR samples (ignoring their different library types)\n* whether we are doing these comparisons before or after batch correction","metadata":{}},{"cell_type":"code","source":"#create upset plots to summarize the overlap between the comparisons performed above\n\n#first create upset plot from the uncorrected data\n#pdf(file=\"Uncorrected-UpSet.pdf\")\nlistInput = list(\"4 UHR Ribo vs 4 HBR Ribo\" = uhr_ribo_vs_hbr_ribo_uncorrected[,\"Gene\"], \n                 \"4 UHR Poly vs 4HBR Poly\" = uhr_poly_vs_hbr_poly_uncorrected[,\"Gene\"],\n                 \"4 UHR Ribo vs 4 HBR Poly\" = uhr_ribo_vs_hbr_poly_uncorrected[,\"Gene\"],\n                 \"4 UHR Poly vs 4 HBR Ribo\" = uhr_poly_vs_hbr_ribo_uncorrected[,\"Gene\"],\n                 \"8 UHR vs 8 HBR\" = uhr_vs_hbr_uncorrected[,\"Gene\"])\nupset(fromList(listInput), order.by = \"freq\", number.angles=45, point.size=3)\n#dev.off()\n\n#now create an upset plot from the batch corrected data\n#pdf(file=\"BatchCorrected-UpSet.pdf\")\nlistInput = list(\"4 UHR Ribo vs 4 HBR Ribo\" = uhr_ribo_vs_hbr_ribo_corrected[,\"Gene\"], \n                 \"4 UHR Poly vs 4 HBR Poly\" = uhr_poly_vs_hbr_poly_corrected[,\"Gene\"],\n                 \"4 UHR Ribo vs 4 HBR Poly\" = uhr_ribo_vs_hbr_poly_corrected[,\"Gene\"],\n                 \"4 UHR Poly vs 4 HBR Ribo\" = uhr_poly_vs_hbr_ribo_corrected[,\"Gene\"],\n                 \"8 UHR vs 8 HBR\" = uhr_vs_hbr_corrected[,\"Gene\"])\nupset(fromList(listInput), order.by = \"freq\", number.angles=45, point.size=3)\n#dev.off()\n\n#write out the final set of DE genes where all UHR and HBR samples were compared using the corrected data\nwrite.table(uhr_vs_hbr_corrected, file=\"DE_genes_uhr_vs_hbr_corrected.tsv\", quote=FALSE, row.names=FALSE, sep=\"\\t\")\n","metadata":{"execution":{"iopub.status.busy":"2022-02-21T07:24:43.742360Z","iopub.execute_input":"2022-02-21T07:24:43.743584Z","iopub.status.idle":"2022-02-21T07:24:45.238584Z"},"trusted":true},"execution_count":17,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}